Covariance

nmheim opened the pull request:
Attempt to make things work with batches of covariance matrices.

Essentially this:
```julia
cov(p::CMeanGaussian{DiagVar}, z::AbstractArray) = map(Diagonal, eachcol(var(p,z)))
cov(p::CMeanGaussian{ScalarVar}, z::AbstractArray) = map(s->I*s, vec(svar(p,z)))

_det(Σ::Diagonal, n::Int) = det(Σ)
_det(Σ::UniformScaling{T}, n::Int) where T = det(Σ*Diagonal(ones(T,n)))

function _logpdf(p::ACGaussian, x::AbstractArray, z::AbstractArray)
    T = eltype(p)
    (μ,Σ) = mean_cov(p,z)
    n = size(μ,1)
    D  = collect(eachcol(x - μ))
    DT = [d' for d in D]
    dΣd = DT .* inv.(Σ) .* D
    -(dΣd .+ log.(_det.(Σ,n)) .+ n*T(log(2π))) / 2
end
```
of course adapted to `CMeanVarGaussian` as necessary.

nmheim committed:
add doc strings

nmheim committed:
changes to variance/logpdf such that stuff works with covariance matrices

vitskvara commented:
Nice. For scalar variances, does this make optimization slower and how much? If it is considerably less performant, maybe we could make an optimized version of logpdf for experiments?

vitskvara reviewed (commented):
Are the changes not gpu compatible? Or did you remove it from the tests just for basic sanity checking?

nmheim commented:
yes, just finishing it on CPU for now, but GPU is hopefully not a big issue. I'll post some timings in a bit, but in theory it should be the same, because the scalar variances are represented by `UniformScaling`

nmheim committed:
CMeanGaussian working on CPU

nmheim committed:
introduce _var_logpdf/_cov_logpdf

all the logpdf functions are currently calling _var_logpdf, because we
are currently only using diagonal covariances anyway.
most of the functionality for _cov_logpdf is there though. there is an
issue with the scalar variance case (broacasting over arrays of
UniformScaling) on GPU though.

nmheim committed:
comment out failing GPU covariance tests for now

nmheim committed:
add gradient tests for all distributions

nmheim commented:
on second thought, I kept the current logpdf implementations. so they are still using normal variance vectors, as we are currently not using full covariances anyway. the functionality is there through `_cov_logpdf`, there is just a one issue on GPU with `CuArray`s of `UniformScaling`s...

nmheim commented:
We are using DistributionsAD now
